juspay/hyperswitch
# feat(multitenancy): add support for multitenancy and handle the same in router, producer, consumer, drainer and analytics

## Type of Change



## Description

1. Add tenant id in sessionState and populate it from request header. All component should take tenant_id only from sessionState
2. Prefix_tenant_id with redis key. Separate connection management for redis is not needed
3. Add redirection url based on the tenant base_url
4. Process tracker to handle tenant-id. To be handled in producer, consumer and drainer

### Additional Changes





## Motivation and Context

 To establish tenancy in our application and start using the default tenant as 'Hyperswitch' and also offer a 'test' tenant in all environments

## How did you test it?


**Test for Scheduler**
1. Enable multitenancy by changing [multitenancy] in development.toml
2. Create multiple tenants by adding [multitenancy.tenants] in development.toml
3. Start Router, Producer and Consumer
4. Create a payment by passing `x-tenant-id` as `<your-tenant>` in the request header in localhost:8080
5. Do the same for both tenants and verify if the payment related data are stored in the respective schema of each tenant
6. Simultaneously each payment should have an process tracker entry in individual schema
7. Verify if the process tracker status for both task is changing to terminal state in DB indicating that producer and consumer are able to pick up the task from different schema and update the status respectively

**Test for Drainer**
1. Enable multitenancy by changing [multitenancy] in development.toml
2. Create multiple tenants by adding [multitenancy.tenants] in development.toml
3. Start Router and Drainer `cargo r --bin drainer`
4. Create a merchant by passing `x-tenant-id` as `<your-tenant>` in the request header in localhost:8080
5. Enable KV for the merchant using below curl
```curl --location 'https://localhost:8080/accounts/{{merchant_id}}/kv' \
--header 'Content-Type: application/json' \
--header 'api-key: {{admin_api_key}}' \
--data '{
  "kv_enabled": true
}'
```
6. Create a payment by passing `x-tenant-id` as `<your-tenant>` in the request header in localhost:8080
7. Once the payment get succeeded, make sure payment_intent and payment_attempt table are updated

**Test for Redirection**
1. Enable multitenancy by changing [multitenancy] in development.toml
2. Create multiple tenants by adding [multitenancy.tenants] in development.toml
3. Start Router and make a 3DS payment, the redirection url should be correct as per the tenant configuration

## Checklist
